src/tests/test_tls_session_manager.cpp

```diff
@@ -0,0 +1,1045 @@
+         auto mock_ticket = Test::rng().random_vec<Botan::TLS::Session_Ticket>(128);
+
+         result.confirm("no session found via ID", !mgr->retrieve(mock_id, cbs, plcy));
+         result.confirm("no session found via ID", !mgr->retrieve(mock_ticket, cbs, plcy));
+         }),
+
+      Botan_Tests::CHECK("clearing empty cache", [&](auto& result)
+         {
+         result.test_eq("does not delete anything", mgr->remove_all(), 0);
+         }),
+
+      Botan_Tests::CHECK("establish new session", [&](auto& result)
+         {
+         auto handle = mgr->establish(default_session(Botan::TLS::Connection_Side::Server, cbs), default_id);
+         if(result.confirm("establishment was successful", handle.has_value()))
+            {
+            result.require("session id was set", handle->id().has_value());
+            result.confirm("session ticket was empty", !handle->ticket().has_value());
+            result.test_is_eq("session id is correct", handle->id().value(), default_id);
+            }
+         }),
+
+      Botan_Tests::CHECK("obtain session from server info", [&](auto& result)
+         {
+         auto sessions = mgr->find(server_info, cbs, plcy);
+         if(result.confirm("session was found successfully", sessions.size() == 1))
+            {
+            result.test_is_eq("protocol version was echoed", sessions[0].first.version(), Botan::TLS::Protocol_Version(Botan::TLS::Version_Code::TLS_V12));
+            result.test_is_eq("ciphersuite was echoed", sessions[0].first.ciphersuite_code(), uint16_t(0x009C));
+            result.test_is_eq("ID was echoed", sessions[0].second.id().value(), default_id);
+            result.confirm("ticket was dropped", !sessions[0].second.ticket().has_value());
+            }
+         }),
+
+      Botan_Tests::CHECK("obtain session from ID", [&](auto& result)
+         {
+         auto session = mgr->retrieve(default_id, cbs, plcy);
+         if(result.confirm("session was found successfully", session.has_value()))
+            {
+            result.test_is_eq("protocol version was echoed", session->version(), Botan::TLS::Protocol_Version(Botan::TLS::Version_Code::TLS_V12));
+            result.test_is_eq("ciphersuite was echoed", session->ciphersuite_code(), uint16_t(0x009C));
+            }
+         }),
+
+      Botan_Tests::CHECK("obtain session from ID disguised as opaque handle", [&](auto& result)
+         {
+         auto session = mgr->retrieve(Botan::TLS::Opaque_Session_Handle(default_id), cbs, plcy);
+         if(result.confirm("session was found successfully", session.has_value()))
+            {
+            result.test_is_eq("protocol version was echoed", session->version(), Botan::TLS::Protocol_Version(Botan::TLS::Version_Code::TLS_V12));
+            result.test_is_eq("ciphersuite was echoed", session->ciphersuite_code(), uint16_t(0x009C));
+            }
+         }),
+
+      Botan_Tests::CHECK("obtain session from ticket == id", [&](auto& result)
```

Comment:
The test description alone sounds like it should work, even if it does in fact not work.

Reply:
Indeed. Thanks. Similar excuse as above. The assumptions changed on the way and the test names didn't keep up.